### DISCUSSION TITLE
Pimcore DB doesn't work as expected

### DISCUSSION THREAD

**Discussion by genfish** [pimcore/pimcore#17165] [Q&A / Support]
Hi!

If I do the following it works fine:

```
$db = Db::get();
$stmt = $db->query($sql);
$row = $stmt->fetch(\PDO::FETCH_ASSOC);
```

But if I do the following, it returns an error the method (fetch, fetchOne, fetchColumn) does not exist:

```
$db = Db::get();
$stmt = $db->prepare($sql);
$stmt ->execute($vars);
$stmt ->fetch(\PDO::FETCH_COLUMN);
```

I can't prepare a SQL and pass the variables to it, so I only have the option of putting them directly into the SQL.

Any idea? Thanks!

**Comment by kingjia90** [comment]
It's not related to Pimcore but to DBAL, you are chaining it wrongly, the [fetch](https://www.doctrine-project.org/projects/doctrine-dbal/en/3.0/reference/data-retrieval-and-manipulation.html#:~:text=If%20you%20are%20finished%20with%20binding%20parameters%20you%20have%20to%20call%20execute()%20on%20the%20statement%2C%20which%20will%20trigger%20a%20query%20to%20the%20database.%20After%20the%20query%20is%20finished%20you%20can%20access%20the%20results%20of%20this%20query%20using%20the%20fetch%20API%20of%20a%20statement) is a method of what `execute()` returns and not of `prepare()`, it should look be something like: 
 ```
$db = Db::get();
$stmt = $db->prepare($sql);
$result = $stmt->execute($vars)->fetch(\PDO::FETCH_COLUMN);
 ```

**Comment by genfish** [comment]
Finaly I found answer, in Pimcore it's not possible pass variables using DB. This is correct method:

```
foreach ($vars as $var)
{
	$value = is_numeric($var) ? $var : $db->quote($var);
	$sql = preg_replace('/\?/', $value, $sql, 1);
}
$rowset = $db->fetchAllAssociative($sql);
```

**Reply by jdreesen** [reply]
No, it's definitely not. It may work, but there are certainly better ways. Please post some more context or code on what you're trying to do.

**Reply by genfish** [reply]
The context of what I'm doing is filtering by multiple fields of the objects associated with a many-to-many field of an objectbrick when there are many objects associated with this field. With Pimcore objects it tries to load too much information and process stops. What other way is there? Thanks for your answer.

**Reply by jdreesen** [reply]
I don't know. I need to see some code.
